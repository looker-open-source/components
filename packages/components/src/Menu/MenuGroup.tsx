/*

 MIT License

 Copyright (c) 2020 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import React, { FC, ReactNode, useContext, useState } from 'react'
import styled from 'styled-components'
import { CompatibleHTMLProps, reset } from '@looker/design-tokens'
import { List } from '../List'
import { MenuItemContext } from './MenuContext'
import { MenuGroupLabel } from './MenuGroupLabel'

export interface MenuGroupProps
  extends Omit<CompatibleHTMLProps<HTMLElement>, 'label'> {
  compact?: boolean
  label?: ReactNode
  className?: string
}

const MenuGroupLayout: FC<MenuGroupProps> = ({
  children,
  className,
  compact,
  label,
}) => {
  const [renderIconPlaceholder, setRenderIconPlaceholder] = useState(false)
  const {
    compact: contextCompact,
    renderIconPlaceholder: contextRenderIconPlaceholder,
    ...restContext
  } = useContext(MenuItemContext)

  const context = {
    ...restContext,
    compact: compact === undefined ? contextCompact : compact,
    renderIconPlaceholder:
      contextRenderIconPlaceholder || renderIconPlaceholder,
    setRenderIconPlaceholder,
  }

  return (
    <li className={className}>
      <MenuItemContext.Provider value={context}>
        {label && <MenuGroupLabel>{label}</MenuGroupLabel>}
        <List>{children}</List>
      </MenuItemContext.Provider>
    </li>
  )
}

export const MenuGroup = styled(MenuGroupLayout)`
  ${reset}
  list-style-type: none;
  padding: ${({ theme: { space } }) => space.xsmall} 0;
`
