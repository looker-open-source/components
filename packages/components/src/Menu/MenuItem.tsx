/*

 MIT License

 Copyright (c) 2020 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import {
  CompatibleHTMLProps,
  size,
  space,
  SizeProps,
  SpaceProps,
} from '@looker/design-tokens'
import { IconNames } from '@looker/icons'
import styled from 'styled-components'
import React, { FC, ReactNode, useContext, useState, useEffect } from 'react'
import { Placement } from '@popperjs/core'
import { Paragraph } from '../Text'
import { useID } from '../utils/useID'
import { Icon } from '../Icon'
import { Tooltip } from '../Tooltip'
import { MenuContext, MenuItemContext } from './MenuContext'
import { MenuItemLayout } from './MenuItemLayout'

export interface MenuItemProps extends CompatibleHTMLProps<HTMLElement> {
  iconArtwork?: ReactNode
  compact?: boolean
  /**
   * Indicates the MenuItem is checked
   */
  current?: boolean
  /*
   * optional extra description
   */
  description?: ReactNode
  detail?: ReactNode
  icon?: IconNames
  /**
   * Sets the correct accessible role for the MenuItem:
   * - Use **'link'** for items that navigation to another page
   * - Use **'button'** for items that trigger in page interactions, like displaying a dialog
   * @default 'button'
   *
   */
  itemRole?: 'link' | 'button'
  tooltip?: string
  tooltipPlacement?: Placement
}

const MenuItemInternal: FC<MenuItemProps> = (props) => {
  const {
    children,
    className,
    compact: propCompact,
    current,
    description,
    detail,
    disabled,
    href,
    icon,
    iconArtwork,
    itemRole,
    onBlur,
    onClick,
    onKeyUp,
    target,
    tooltip,
    tooltipPlacement,
  } = props

  const [isFocusVisible, setFocusVisible] = useState(false)
  const { compact: contextCompact } = useContext(MenuItemContext)
  const compact = propCompact === undefined ? contextCompact : propCompact

  const handleOnBlur = (event: React.FocusEvent<HTMLLIElement>) => {
    setFocusVisible(false)
    onBlur && onBlur(event)
  }

  const { setOpen } = useContext(MenuContext)
  const {
    renderIconPlaceholder,
    setRenderIconPlaceholder,
    handleArrowDown,
    handleArrowUp,
  } = useContext(MenuItemContext)

  const handleOnClick = (event: React.MouseEvent<HTMLLIElement>) => {
    setFocusVisible(false)
    onClick && onClick(event)
    // Close the Menu (unless event has preventDefault in onClick)
    if (setOpen && !event.defaultPrevented) {
      setOpen(false)
    }
  }

  const handleOnKeyDown = (event: React.KeyboardEvent<HTMLLIElement>) => {
    switch (event.key) {
      case 'ArrowUp':
        handleArrowUp && handleArrowUp(event)
        break
      case 'ArrowDown':
        handleArrowDown && handleArrowDown(event)
        break
    }
  }

  const handleOnKeyUp = (event: React.KeyboardEvent<HTMLLIElement>) => {
    onKeyUp && onKeyUp(event)
    setFocusVisible(true)
  }

  useEffect(() => {
    icon && setRenderIconPlaceholder(true)
  }, [icon, setRenderIconPlaceholder])

  const renderedIconID = useID(props.id)

  const renderedIcon =
    icon || iconArtwork ? (
      <Icon
        artwork={iconArtwork}
        color="text1"
        name={icon}
        size={24 / (compact ? 1.25 : 1)}
        mr="xsmall"
      />
    ) : (
      renderIconPlaceholder && (
        <IconPlaceholder
          aria-hidden
          data-testid={`menu-item-${renderedIconID}-icon-placeholder`}
          mr="xsmall"
          size={24 / (compact ? 1.25 : 1)}
        />
      )
    )

  if (disabled && itemRole === 'link') {
    // eslint-disable-next-line no-console
    console.warn(
      'itemRole="link" and disabled cannot be combined - use itemRole="button" if you need to offer a disabled MenuItem'
    )
  }
  const Component = !disabled && itemRole === 'link' ? 'a' : 'button'

  const menuItemContent = (
    <Component href={href} role="menuitem" target={target}>
      {renderedIcon}
      <span>
        {children}
        {description && (
          <Paragraph color="text2" fontSize="xsmall" mt="xxsmall">
            {description}
          </Paragraph>
        )}
      </span>
      {detail && <Detail>{detail}</Detail>}
    </Component>
  )

  return (
    <MenuItemLayout
      aria-current={current && 'true'}
      compact={compact}
      disabled={disabled}
      focusVisible={isFocusVisible}
      onBlur={handleOnBlur}
      onClick={disabled ? undefined : handleOnClick}
      onKeyUp={handleOnKeyUp}
      onKeyDown={handleOnKeyDown}
      className={className}
    >
      {tooltip ? (
        <Tooltip placement={tooltipPlacement || 'left'} content={tooltip}>
          {menuItemContent}
        </Tooltip>
      ) : (
        menuItemContent
      )}
    </MenuItemLayout>
  )
}

const Detail = styled.div`
  color: ${({ theme: { colors } }) => colors.text1};
  margin-left: auto;
  margin-right: ${({ theme: { space } }) => space.medium};
  padding-left: ${({ theme: { space } }) => space.large};
`

export const MenuItem = styled(MenuItemInternal)``

interface IconPlaceholderProps extends SizeProps, SpaceProps {}

const IconPlaceholder = styled.div<IconPlaceholderProps>`
  ${size}
  ${space}
`
