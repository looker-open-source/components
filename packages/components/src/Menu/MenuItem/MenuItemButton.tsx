/*

 MIT License

 Copyright (c) 2020 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import styled from 'styled-components'
import {
  reset,
  space,
  CompatibleHTMLProps,
  SpaceProps,
} from '@looker/design-tokens'

export interface MenuItemButtonProps
  extends SpaceProps,
    CompatibleHTMLProps<HTMLElement> {}

export const MenuItemButton = styled.button<MenuItemButtonProps>`
  ${reset}
  ${space}

  align-items: center;
  background: transparent;
  border: 0;
  color: inherit;
  cursor: pointer;
  display: flex;
  font-size: inherit;
  flex: 1;
  outline: none;
  text-align: left;
  text-decoration: none;

  &:hover,
  &:focus {
    text-decoration: none;
    color: inherit;
  }
`
