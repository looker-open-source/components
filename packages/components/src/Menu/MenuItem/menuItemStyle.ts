/*

 MIT License

 Copyright (c) 2020 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import { FontWeights, FontSizes, theme } from '@looker/design-tokens'

const { colors } = theme

export interface MenuItemStateStyle {
  bg: string
  color: string
  fontWeight: FontWeights
  fontSize: FontSizes
  iconColor: string
  iconSize: number
}

export interface MenuItemStyle {
  current: MenuItemStateStyle
  hover: MenuItemStateStyle
  initial: MenuItemStateStyle
  marker: {
    color: string
    size: number
  }
}

const initial: MenuItemStateStyle = {
  bg: colors.background,
  color: colors.text3,
  fontSize: 'small',
  fontWeight: 'normal',
  iconColor: colors.text6,
  iconSize: 20,
}

const hover: MenuItemStateStyle = {
  ...initial,
  bg: colors.ui1,
  color: colors.text0,
}

const current: MenuItemStateStyle = {
  ...hover,
  fontWeight: 'bold',
  iconColor: colors.text0,
}

export const defaultMenuItemStyle: MenuItemStyle = {
  current,
  hover,
  initial,
  marker: {
    color: colors.text0,
    size: 4,
  },
}
