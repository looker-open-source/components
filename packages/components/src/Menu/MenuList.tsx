/*

 MIT License

 Copyright (c) 2020 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import { Placement } from '@popperjs/core'
import React, { Ref, useRef, forwardRef, useContext, useState } from 'react'
import { HotKeys } from 'react-hotkeys'
import styled, { css } from 'styled-components'
import {
  MaxHeightProps,
  MinHeightProps,
  HeightProps,
  height,
  minHeight,
  maxHeight,
  width,
  WidthProps,
  MaxWidthProps,
  MinWidthProps,
  minWidth,
  maxWidth,
} from 'styled-system'
import { CompatibleHTMLProps, reset } from '@looker/design-tokens'
import { usePopover } from '../Popover'
import { MenuContext, MenuItemStyleContext } from './MenuContext'
import { MenuGroup } from './MenuGroup'
import { MenuSharedProps } from './MenuItem'
import { moveFocus } from './moveFocus'

export interface MenuListProps
  extends CompatibleHTMLProps<HTMLUListElement>,
    MaxHeightProps,
    MinHeightProps,
    HeightProps,
    MaxWidthProps,
    MinWidthProps,
    WidthProps,
    MenuSharedProps {
  compact?: boolean
  groupDividers?: boolean

  /**
   * Can be one of: top, bottom, left, right, auto, with the modifiers: start,
   * end. This value comes directly from popper.js. See
   * https://popper.js.org/popper-documentation.html#Popper.placements for more
   * info.
   * @default bottom
   */
  placement?: Placement

  /**
   * By default MenuList will reposition itself if they overflow the widow.
   * You can use the pin property to override this behavior.
   */
  pin?: boolean

  /**
   * Allow the overlay to break out of the scroll parent
   */
  escapeWithReference?: boolean
}

export const MenuListInternal = forwardRef(
  (
    {
      children,
      compact,
      customizationProps,
      disabled,
      pin,
      placement,
      ...props
    }: MenuListProps,
    ref: Ref<HTMLUListElement>
  ) => {
    const { id, isOpen, setOpen, triggerElement } = useContext(MenuContext)

    const [renderIconPlaceholder, setRenderIconPlaceholder] = useState(false)

    const innerRef = useRef<null | HTMLElement>(null)

    const context = {
      compact,
      customizationProps,
      renderIconPlaceholder,
      setRenderIconPlaceholder,
    }

    const menuList = (
      <MenuItemStyleContext.Provider value={context}>
        <HotKeys
          innerRef={innerRef}
          keyMap={{ MOVE_DOWN: 'down', MOVE_UP: 'up' }}
          handlers={{
            MOVE_DOWN: () => moveFocus(1, 0, innerRef),
            MOVE_UP: () => moveFocus(-1, -1, innerRef),
          }}
          style={{ borderRadius: 'inherit' }}
        >
          <ul
            ref={ref}
            tabIndex={-1}
            role="menu"
            id={id}
            aria-labelledby={id && `button-${id}`}
            {...props}
          >
            {children}
          </ul>
        </HotKeys>
      </MenuItemStyleContext.Provider>
    )

    const isMenu = isOpen !== undefined
    const { popover } = usePopover({
      content: menuList,
      isOpen,
      pin,
      placement,
      setOpen,
      triggerElement,
    })

    if (disabled) return null

    if (isMenu) return popover || null

    return menuList
  }
)

const dividersStyle = css`
  ${MenuGroup} ~ ${MenuGroup} { /* stylelint-disable-line */
    border-top: 1px solid ${(props) => props.theme.colors.palette.charcoal200};
  }
`

export const MenuList = styled(MenuListInternal)`
  ${reset}

  ${minHeight}
  ${maxHeight}
  ${height}

  min-width: 12rem;
  ${minWidth}
  ${maxWidth}
  ${width}

  overflow: auto;
  border-radius: inherit;

  list-style: none;
  outline: none;
  user-select: none;
  ${(props) => props.groupDividers !== false && dividersStyle};
`
