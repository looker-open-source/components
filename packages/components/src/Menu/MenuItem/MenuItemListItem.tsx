/*

 MIT License

 Copyright (c) 2020 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import omit from 'lodash/omit'
import {
  CompatibleHTMLProps,
  color,
  space,
  typography,
} from '@looker/design-tokens'
import React, { forwardRef, Ref } from 'react'
import styled, { css } from 'styled-components'
import { Icon } from '../../Icon'
import { MenuItemStyle } from './menuItemStyle'

export interface MenuListItemProps extends CompatibleHTMLProps<HTMLLIElement> {
  current?: boolean
  focusVisible?: boolean
  itemStyle: MenuItemStyle
}

const hoverStyles = (props: MenuListItemProps) => {
  if (props.current) return false

  return css`
    &:hover {
      background: ${props.itemStyle.hover.bg};
      color: ${props.itemStyle.hover.color};
      font-size: ${props.itemStyle.hover.color};
      font-weight: ${props.itemStyle.hover.color};

      ${Icon} {
        color: ${props.itemStyle.hover.iconColor};
      }
    }
  `
}

const iconColor = (props: MenuListItemProps) =>
  props.current
    ? props.itemStyle.current.iconColor
    : props.itemStyle.initial.iconColor

/**
 * All of this drama is to deal with SC's behavior of auto-spreading the Element interface
 * used when styled extends a base type. E.g. (styled.li has `color` prop)
 */
const Li = forwardRef((props: MenuListItemProps, ref: Ref<HTMLLIElement>) => {
  const domProps = omit(props, 'current', 'focusVisible', 'itemStyle')
  return <li {...domProps} ref={ref} />
})

Li.displayName = 'Li'

export const MenuItemListItem = styled(Li)<MenuListItemProps>`
  ${color}
  ${space}
  ${typography}

  align-items: center;
  display: flex;
  flex-wrap: wrap;
  outline: none;
  text-decoration: none;
  transition: ${(props) =>
    `background ${props.theme.transitions.durationQuick} ${props.theme.easings.ease},
    color ${props.theme.transitions.durationQuick} ${props.theme.easings.ease}`};

  button,
  a {
    border-left-width: ${({ itemStyle }) => itemStyle.marker.size}px;
    border-left-style: solid;
    border-left-color: ${({ itemStyle, current }) =>
      current ? itemStyle.marker.color : 'transparent'};
    outline: none;
    padding-left: calc(${({ theme, itemStyle }) =>
      `${theme.space.medium} - ${itemStyle.marker.size}px`});
  }

  ${(props) =>
    props.focusVisible &&
    `&:focus-within button,
    &:focus-within a {
      box-shadow: 0 0 3px 1px ${props.theme.colors.key};
    }
`}

  ${hoverStyles};

  ${Icon} {
    color: ${iconColor};
    transition: color
      ${(props) =>
        `${props.theme.transitions.durationQuick} ${props.theme.easings.ease}`};
  }
`
