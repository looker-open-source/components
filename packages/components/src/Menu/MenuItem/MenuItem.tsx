/*

 MIT License

 Copyright (c) 2020 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import merge from 'lodash/merge'
import {
  CompatibleHTMLProps,
  ColorProps,
  FontSizes,
  SpaceProps,
  TypographyProps,
} from '@looker/design-tokens'
import { IconNames } from '@looker/icons'
import React, { FC, ReactNode, useContext, useState } from 'react'
import { Icon } from '../../Icon'
import { MenuItemStyleContext } from '../MenuContext'
import { MenuItemButton } from './MenuItemButton'
import {
  MenuItemCustomization,
  MenuItemStateCustomizations,
} from './menuItemCustomization'
import { MenuItemDetail } from './MenuItemDetail'
import { MenuItemListItem } from './MenuItemListItem'
import {
  defaultMenuItemStyle,
  MenuItemStateStyle,
  MenuItemStyle,
} from './menuItemStyle'

export interface MenuSharedProps {
  customizationProps?: MenuItemCustomization
  compact?: boolean
}

// For merging compact and customizationProps from props with those from context
export function useMenuItemStyleContext(props: MenuSharedProps) {
  const { customizationProps: propCustomizations, compact: compactProp } = props
  const {
    customizationProps: contextCustomizations,
    compact: contextCompact,
  } = useContext(MenuItemStyleContext)
  const parentCustomizations = contextCustomizations || {}

  let customizationProps = parentCustomizations || propCustomizations
  if (customizationProps && parentCustomizations) {
    customizationProps = merge({}, parentCustomizations, propCustomizations)
  }

  const compact = compactProp === undefined ? contextCompact : compactProp

  return { compact, customizationProps }
}

const assignCustomizations = (
  defaultStyle: MenuItemStyle,
  changes?: MenuItemCustomization
): MenuItemStyle => {
  const { bg, color, iconColor, fontWeight, fontSize, iconSize } =
    changes || ({} as MenuItemCustomization)

  const customMarker = changes ? changes.marker : {}

  // Need to spread fontSize & iconSize across all states
  const defaults = {
    fontSize: fontSize || ('small' as FontSizes),
    iconSize: iconSize || 20,
  }

  const base: MenuItemStateCustomizations = {}

  bg && (base.bg = bg)
  color && (base.color = color)
  fontWeight && (base.fontWeight = fontWeight)
  iconColor && (base.iconColor = iconColor)

  const initial: MenuItemStateStyle = {
    ...defaultStyle.initial,
    ...defaults,
    ...base,
  }

  const current: MenuItemStateStyle = {
    ...defaultStyle.current,
    ...defaults,
    ...base,
    ...(changes ? changes.current : {}),
  }

  const hover: MenuItemStateStyle = {
    ...defaultStyle.hover,
    ...defaults,
    ...base,
    ...(changes ? changes.hover : {}),
  }

  const marker = { ...defaultStyle.marker, ...customMarker }

  return {
    current,
    hover,
    initial,
    marker,
  }
}

export interface MenuItemProps
  extends CompatibleHTMLProps<HTMLElement>,
    ColorProps,
    SpaceProps,
    TypographyProps,
    MenuSharedProps {
  detail?: ReactNode
  icon?: IconNames
  /**
   * Indicates the MenuItem is checked
   */
  current?: boolean
  /**
   * Sets the correct accessible role for the MenuItem:
   * - Use **'link'** for items that navigation to another page
   * - Use **'button'** for items that trigger in page interactions, like displaying a modal
   * @default 'button'
   *
   */
  itemRole?: 'link' | 'button'
}

export const MenuItem: FC<MenuItemProps> = (props) => {
  const {
    children,
    compact: propCompact,
    customizationProps: propCustomizations,
    current,
    detail,
    href,
    icon,
    itemRole,
    onBlur,
    onClick,
    onKeyUp,
    target,
    ...remainingProps
  } = props

  const [isFocusVisible, setFocusVisible] = useState(false)

  const handleOnKeyUp = (event: React.KeyboardEvent<HTMLLIElement>) => {
    setFocusVisible(true)
    onKeyUp && onKeyUp(event)
  }

  const handleOnBlur = (event: React.FocusEvent<HTMLLIElement>) => {
    setFocusVisible(false)
    onBlur && onBlur(event)
  }

  const handleOnClick = (event: React.MouseEvent<HTMLLIElement>) => {
    setFocusVisible(false)
    onClick && onClick(event)
  }

  const { customizationProps, compact } = useMenuItemStyleContext({
    compact: propCompact,
    customizationProps: propCustomizations,
  })
  const compactIconModifier = compact ? 1.25 : 1

  const style = assignCustomizations(defaultMenuItemStyle, customizationProps)
  const styleState = current ? style.current : style.initial
  const { iconSize, iconColor, ...listItemProps } = styleState

  const { p, py, px, pr, pl, pt, pb, ...outerProps } = remainingProps

  const clickTargetProps = {
    p,
    pb: pb || py || p || compact ? 'xxsmall' : 'small',
    pl: pl || px || p || 'medium',
    pr: pr || px || p || 'medium',
    pt: pt || py || p || compact ? 'xxsmall' : 'small',
  }

  return (
    <MenuItemListItem
      aria-current={current && 'page'}
      onClick={handleOnClick}
      current={current}
      focusVisible={isFocusVisible}
      itemStyle={style}
      onKeyUp={handleOnKeyUp}
      onBlur={handleOnBlur}
      {...listItemProps}
      {...outerProps}
    >
      <MenuItemButton
        as={itemRole === 'link' ? 'a' : 'button'}
        href={href}
        role="menuitem"
        target={target}
        {...clickTargetProps}
      >
        {icon && (
          <Icon
            name={icon}
            mr="xsmall"
            size={iconSize / compactIconModifier}
            color={iconColor}
          />
        )}
        {children}
      </MenuItemButton>
      {detail && <MenuItemDetail>{detail}</MenuItemDetail>}
    </MenuItemListItem>
  )
}
