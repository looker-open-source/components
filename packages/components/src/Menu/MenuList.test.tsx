/*

 MIT License

 Copyright (c) 2020 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import { screen } from '@testing-library/react'
import 'jest-styled-components'
import * as React from 'react'
import { renderWithTheme } from '@looker/components-test-utils'
import { MenuGroup } from './MenuGroup'
import { MenuItem } from './MenuItem'
import { MenuList } from './MenuList'

/* eslint-disable-next-line @typescript-eslint/unbound-method */
const globalGetBoundingClientRect = Element.prototype.getBoundingClientRect

describe('MenuList', () => {
  test('allocates space for MenuItem when a sibling has an icon', () => {
    const { getByTestId } = renderWithTheme(
      <MenuList>
        <MenuItem icon="Calendar">Gouda</MenuItem>
        <MenuItem id="cheddar">Cheddar</MenuItem>
      </MenuList>
    )

    getByTestId('menu-item-cheddar-icon-placeholder')
  })

  describe('windowing', () => {
    beforeEach(() => {
      jest.useFakeTimers()
      /* eslint-disable-next-line @typescript-eslint/unbound-method */
      Element.prototype.getBoundingClientRect = jest.fn(() => {
        return {
          bottom: 0,
          height: 342,
          left: 0,
          right: 0,
          toJSON: jest.fn(),
          top: 0,
          width: 260,
          x: 0,
          y: 0,
        }
      })
    })

    afterEach(() => {
      jest.runOnlyPendingTimers()
      jest.useRealTimers()
      jest.resetAllMocks()
      /* eslint-disable-next-line @typescript-eslint/unbound-method */
      Element.prototype.getBoundingClientRect = globalGetBoundingClientRect
    })

    test('fixed', () => {
      const arr3000 = Array.from(Array(3000), (_, i) => i)
      renderWithTheme(
        <MenuList>
          {arr3000.map((num) => (
            <MenuItem key={num}>{num}</MenuItem>
          ))}
        </MenuList>
      )

      expect(screen.getByText('0')).toBeVisible()
      expect(screen.getByText('14')).toBeVisible()
      expect(screen.queryByText('15')).not.toBeInTheDocument()

      expect(screen.queryByTestId('before')).not.toBeInTheDocument()
      expect(screen.getByTestId('after')).toHaveStyle('height: 119400px;')
    })

    test('variable', () => {
      const arr200 = Array.from(Array(200), (_, i) => i)
      renderWithTheme(
        <MenuList>
          {arr200.map((num) => (
            <MenuGroup key={num}>
              {Array.from(Array((num + 1) % 15), (_, i) => (
                <MenuItem key={i}>{`${num}_${i}`}</MenuItem>
              ))}
            </MenuGroup>
          ))}
        </MenuList>
      )

      expect(screen.getByText('0_0')).toBeVisible()
      expect(screen.getByText('6_6')).toBeVisible()
      expect(screen.queryByText('7_0')).not.toBeInTheDocument()

      expect(screen.queryByTestId('before')).not.toBeInTheDocument()
      expect(screen.getByTestId('after')).toHaveStyle('height: 61993px;')
    })
  })
})
