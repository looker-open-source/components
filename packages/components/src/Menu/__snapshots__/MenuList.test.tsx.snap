// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu - compact 1`] = `ReactWrapper {}`;

exports[`Menu - composed 1`] = `ShallowWrapper {}`;

exports[`Menu 1`] = `ShallowWrapper {}`;
